x-pack/plugins/apm/public/components/app/transaction_details/WaterfallWithSummmary/WaterfallContainer/Waterfall/waterfall_helpers/waterfall_helpers.ts

```diff
@@ -203,27 +225,39 @@ function getRootTransaction(childrenByParentId: IWaterfallGroup) {
   }
 }
 
-export type IServiceColors = Record<string, string>;
+function getLegends(waterfallItems: IWaterfallItem[]) {
+  const onlySpanItems = waterfallItems.filter(
+    (item) => 'legendValues' in item
+  ) as IWaterfallSpanItem[];
+
+  const legends = [
+    WaterfallLegendType.ServiceName,
+    WaterfallLegendType.SpanType,
+  ].flatMap((legendType) => {
+    const values = uniq(
+      onlySpanItems.map((item) => item.legendValues[legendType])
+    );
 
-function getServiceColors(waterfallItems: IWaterfallItem[]) {
-  const services = uniq(waterfallItems.map((item) => item.doc.service.name));
+    const palette = euiPaletteColorBlind({
+      rotations: Math.ceil(values.length / 10),
+    });
 
-  const assignedColors = [
-    theme.euiColorVis1,
-    theme.euiColorVis0,
-    theme.euiColorVis3,
-    theme.euiColorVis2,
-    theme.euiColorVis6,
-    theme.euiColorVis7,
-    theme.euiColorVis5,
-  ];
+    return values.map((value, index) => ({
+      type: legendType,
+      value,
```

Comment:
```suggestion
      label,
```

Reply:
It's a value (it can be undefined as well). We convert to label elsewhere (in `WaterfallLegends.tsx`).